app/src/main/java/org/dslul/openboard/inputmethod/latin/settings/UserDictionaryAddWordContents.java

```diff
@@ -0,0 +1,308 @@
+        if (TextUtils.isEmpty(mSavedWeight)) {
+            mSavedWeight = String.valueOf(WEIGHT_FOR_USER_DICTIONARY_ADDS);
+        }
+
+        mSavedWord = newWord;
+
+        // In edit mode, everything is modified without overwriting other existing words
+        if (MODE_EDIT == mMode && hasWord(newWord, locale, context) && newWord.equals(mOldWord)) {
+            UserDictionarySettings.deleteWordInEditMode(mOldWord, mOldShortcut, mOldWeight, locale, resolver);
+            // Toast appears to indicate that the word has been modified
+            Toast.makeText(context, context.getText(R.string.user_dict_word_modified), Toast.LENGTH_SHORT).show();
+        } else {
+            mMode = MODE_INSERT;
+        }
+
+        if (mMode == MODE_INSERT && hasWord(newWord, locale, context)) {
+            return CODE_ALREADY_PRESENT;
+        }
+
+        if (mMode == MODE_INSERT) {
+            // Delete duplicate when adding or updating new word
+            UserDictionarySettings.deleteWordInEditMode(mOldWord, mOldShortcut, mOldWeight, locale, resolver);
+            // Update the existing word by adding a new one
+            UserDictionary.Words.addWord(context, newWord,
+                    Integer.parseInt(mSavedWeight), mSavedShortcut, TextUtils.isEmpty(mLocaleString) ?
+                            null : LocaleUtils.constructLocaleFromString(mLocaleString));
+
+            // Toast appears either to indicate that the word has been modified or created
+            if (!TextUtils.isEmpty(mOldWord)) {
+                Toast.makeText(context, context.getText(R.string.user_dict_word_modified), Toast.LENGTH_SHORT).show();
+            } else {
+                final String localeDisplayName = UserDictionarySettings.getLocaleDisplayName(context, mLocaleString);
+                final String messageWordAdded = context.getString(R.string.user_dict_word_added, localeDisplayName);
+                Toast.makeText(context, messageWordAdded, Toast.LENGTH_SHORT).show();
+            }
+            return CODE_UPDATED;
+        }
+
+        // Delete duplicates
+        UserDictionarySettings.deleteWord(newWord, locale, resolver);
+
+        // In this class we use the empty string to represent 'all locales' and mLocale cannot
+        // be null. However the addWord method takes null to mean 'all locales'.
+        UserDictionary.Words.addWord(context, newWord,
+                Integer.parseInt(mSavedWeight), mSavedShortcut, TextUtils.isEmpty(mLocaleString) ?
+                        null : LocaleUtils.constructLocaleFromString(mLocaleString));
+
+        return CODE_WORD_ADDED;
+    }
+
+    public boolean isExistingWord(final Context context) {
+        final String newWord = mWordEditText.getText().toString();
+        if (mMode == MODE_INSERT || apply(context) == CODE_ALREADY_PRESENT) {
+            return hasWord(newWord, mLocaleString, context);
+        } else {
+            return false;
+        }
```

Comment:
This might be working, but is semantically very strange and seems to have side effects.
When you call a function `isExistingWord`, a call to `apply` is definitely unexpected.

Could it just always call `hasWord`?

Reply:
I can't remember why I wrote it like that but there's definitely a reason; I think it's because of the editing mode. I'll have to do some tests to remember exactly.

Reply:
That's what I thought: I wrote the code like that to prevent the dictionary from considering that the word already exists when we modify its data (the shortcut or the weight).

Reply:
If we write

```java
if (mMode != MODE_EDIT) {
```
 it should be ok according to my tests. I'll make the change

Reply:
Done with commit c3d5935